node/testing/src/simulator/mod.rs

```diff
@@ -217,19 +220,42 @@ impl Simulator {
         self.wait_for_all_nodes_synced(runner).await;
     }
 
+    pub async fn setup_and_run_with_listener<'a, AL, ALF>(
+        &mut self,
+        runner: &mut ClusterRunner<'a>,
+        listener: ALF,
+    ) where
+        ALF: FnMut() -> AL,
+        AL: 'static
+            + Send
+            + FnMut(ClusterNodeId, &State, &NodeTestingService, &ActionWithMeta) -> bool,
+    {
+        self.setup(runner).await;
+        self.run_with_listener(runner, listener).await;
+    }
+
+    pub async fn setup_and_run<'a>(&mut self, runner: &mut ClusterRunner<'a>) {
+        self.setup(runner).await;
+        self.run_with_listener(runner, || |_, _, _, _| false).await;
+    }
+
     pub async fn setup<'a>(&mut self, runner: &mut ClusterRunner<'a>) {
         self.set_up_seed_nodes(runner).await;
         self.set_up_normal_nodes(runner).await;
         self.set_up_snark_worker_nodes(runner).await;
         self.set_up_block_producer_nodes(runner).await;
     }
 
-    pub async fn setup_and_run<'a>(&mut self, runner: &mut ClusterRunner<'a>) {
-        self.setup(runner).await;
-        self.run(runner).await;
-    }
-
-    pub async fn run<'a>(&mut self, runner: &mut ClusterRunner<'a>) {
+    pub async fn run_with_listener<'a, AL, ALF>(
```

Comment:
please bring back `run` as well and use that for testing/src/server/simulator.rs

Reply:
done